pkg/statsd/handler_tags.go

```diff
@@ -34,7 +54,52 @@ func (th *TagHandler) DispatchMetric(ctx context.Context, m *gostatsd.Metric) er
 		m.Hostname = string(m.SourceIP)
 	}
 	m.Tags = uniqueTags(m.Tags, th.tags)
-	return th.metrics.DispatchMetric(ctx, m)
+	if th.filterMetricAndTags(m) {
+		return th.metrics.DispatchMetric(ctx, m)
+	}
+	return nil
+}
+
+// returns true if the metric should be submitted to the next stage.  Updates the metric tags if needed.
+func (th *TagHandler) filterMetricAndTags(m *gostatsd.Metric) bool {
+	for _, filter := range th.filters {
+		if len(filter.MatchMetrics) > 0 && !filter.MatchMetrics.MatchAny(m.Name) { // returns false if nothing present
+			// name doesn't match an include, stop
+			continue
+		}
+
+		// this list may be empty, and therefore return false
+		if filter.ExcludeMetrics.MatchAny(m.Name) { // returns false if nothing present
+			// name matches an exclude, stop
+			continue
+		}
+
+		if len(filter.MatchTags) > 0 && !filter.MatchTags.MatchAnyMultiple(m.Tags) { // returns false if either list is empty
+			// no tags match
+			continue
+		}
+
+		if filter.DropMetric {
+			return false
+		}
+
+		// TODO: This is not the most efficient.  It might be better to replace with an empty string, and ensure empty
+		// tags are handled downstream.
```

Comment:
Current implementation does _n_ copies for _m_ filters, or O(n*m) copies. If you iterate over the tags and then test each filter in turn you can make this O(n) copies. If we don't care about preserving order then we can reduce that further to just the number of dropped tags, which may be zero a lot of the time.

Reply:
Yeah, it's pretty bad.  I'll collect the tags to drop and pass it to ```uniqueTags```.  It needs to build a list regardless.

Reply:
It doesn't need to do _n_ copies though. We're destructively appending to a slice backed by the same array. You can delete in place by replacing m.Tags[i] with m.Tags[len(m.Tags)-1] and resize the slice.

    a[i] = a[len(a)-1] 
    a = a[:len(a)-1]

https://github.com/golang/go/wiki/SliceTricks

Reply:
Ahh, got it.  I can incorporate that.